src/Hazelcast.Net/Sql/SqlResult.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Hazelcast.Core;
+using Hazelcast.Serialization;
+
+namespace Hazelcast.Sql
+{
+    // FIXME [Oleksii] Clarify if should be thread safe
+    public class SqlResult : IAsyncEnumerator<SqlRow>
```

Comment:
I didn't make `SqlResult` implement `IAsyncEnumerable`, but `IAsyncEnumerator` instead. The reason is that it doesn't fully match the idea behind enumerable, which usually can be iterated over any number of times via calling `GetEnumerator` . Unfortunately, `SqlResult` can only be iterated once. To highlight this behavior for users, I implemented it as `IAsyncEnumerator` and added `EnumerateOnce` method that will return one-time use `IAsyncEnumerable`.

Reply:
@zpqrtbnk I think `EnumerateOnce` make if more straightforward that result is a one-time use object. In most cases if developer sees an `IEnumerable` and haven't yet checked the documentation he assumes it can be enumerated until it's cleaned up or disposed.

Reply:
Will go with `EnumerateOnce` for now but leave this for open for discussion for now and/or possible change in future versions.

Reply:
`EnumerateOnce` is removed in favor of `IAsyncEnumerable` and System.Linq.Async package.